digitiser_decode.cpp

```diff
@@ -69,6 +71,126 @@ static std::vector<std::int16_t> decode_10bit(const std::uint8_t *data, std::siz
     return out;
 }
 
+template<int left>
+[[gnu::target("avx2")]]
+static inline __m256i extr(__m256i in)
+{
+    static_assert(0 <= left && left <= 22, "left is out of ange");
+    if (left)
+        return _mm256_srai_epi32(_mm256_slli_epi32(in, left), 22);
```

Comment:
Why use arithmetic shifting and not just logical shifting throughout? Is it not just about getting the bit pattern in the right spot?

Reply:
The digitiser samples are signed, so the right shift needs to be arithmetic to do sign extension (the intrinsic name for left shift is "sll", but there is only one kind of left shift).